### PR TITLE
Add path to save plot and some improvement in draw a figure

### PR DESCRIPTION
**Pull request by imankahbasi**
### What does this PR do?
add feature to save figure of plot in a directory
improve shape of figure
### PR Checklist
- [x] `develop` base branch selected?
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests still pass.
- [ ] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGES.md`.
~~- [ ] First time contributors have added your name to `CONTRIBUTORS.md`.~~

### CODE REVIEW

File: eqcorrscan/utils/plotting.py [hunk 1]
```diff
@@ -1172,7 +1171,7 @@ def threeD_seismplot(stations, nodes, size=(10.5, 7.5), **kwargs):
     return fig
 
 
-def noise_plot(signal, noise, normalise=False, **kwargs):
+def noise_plot(signal, noise, normalise=False, plotdir=None, **kwargs):
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/plotting.py]: Please remove `plotdir` argument, it should be handled by the `savefile` kwarg.

File: eqcorrscan/utils/plotting.py [hunk 2]
```diff
@@ -1246,17 +1248,32 @@ def noise_plot(signal, noise, normalise=False, **kwargs):
     plt.figlegend(lines, labels, 'upper left')
     plt.tight_layout()
     plt.subplots_adjust(hspace=0)
-    fig = _finalise_figure(fig=fig, **kwargs)  # pragma: no cover
-    return fig
+    plt.subplots_adjust(top=0.91)
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/plotting.py]: You should incorporate this in the above subplots adjust if you want it.  It is likely to break our image comparison tests though, so if its not needed then avoid it.

File: eqcorrscan/utils/plotting.py [hunk 3]
```diff
@@ -1246,17 +1248,32 @@ def noise_plot(signal, noise, normalise=False, **kwargs):
...
+    if plotdir:
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/plotting.py]: This logic should be in the calling function, see earlier comment.  savefile, title, etc should be handled by kwargs.

File: eqcorrscan/utils/plotting.py [hunk 4]
```diff
@@ -1246,17 +1248,32 @@ def noise_plot(signal, noise, normalise=False, **kwargs):
...
+    else:
+        _finalise_figure(
+            fig=fig, show=True,
+            title='SNR for ' + str(signal[0].stats.starttime))
+        # pragma: no cover
 
 
-def pretty_template_plot(template, size=(10.5, 7.5), background=False,
-                         picks=False, **kwargs):
+def pretty_template_plot(template, plotdir=None, size=(10.5, 7.5),
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/plotting.py]: Please remove plotdir, this should be handled by kwargs

File: eqcorrscan/core/template_gen.py [hunk 5]
```diff
@@ -807,15 +815,13 @@ def _template_gen(picks, st, length, swin='all', prepick=0.05,
...
-        fig1.savefig("{0}_template.png".format(st1[0].stats.starttime))
-        fig2.savefig("{0}_noise.png".format(st1[0].stats.starttime))
-        del(stplot, fig1, fig2)
+    if plot and plotdir and len(st1) > 0:
+        tplot(st1, background=stplot, picks=picks_copy, plotdir=plotdir)
+        noise_plot(signal=st1, noise=noise, plotdir=plotdir)
+    elif plot and len(st1) > 0:
+        tplot(st1, background=stplot, picks=picks_copy, show=True)
+        noise_plot(signal=st1, noise=noise, show=True)
+    del(stplot)
```
Comment (calum-chamberlain) [inline eqcorrscan/core/template_gen.py]: This breaks the tests - `stplot` only exists if `plot=True`. Can you change this to something like:
```python
if plot:
    plot_kwargs = dict(show=True, save=False)
    if plotdir:
        if not os.path.isdir(plotdir):
            os.makedirs(plotdir)
        plot_kwargs.update({"save": True, "show": False})
    plot_kwargs.update(
        {"savefile": "{0}/{1}_SNR.png".format(plotdir, st1[0].stats.starttime),
         "title": "SNR for {0}".format(st1[0].stats.starttime)})
    noise_plot(signal=st1, noise=noise, **plot_kwargs)
    plot_kwargs.update(
        {"savefile": "{0}/{1_template.png}".format(plotdir, template[0].stats.starttime),
         "title": "SNR for {0}".format(st1[0].stats.starttime)})
    tplot(st1, background=stplot, picks=picks_copy, **plot_kwargs)
    del(stplot)
```

File: eqcorrscan/utils/plotting.py [hunk 6]
```diff
@@ -1398,19 +1415,22 @@ def pretty_template_plot(template, size=(10.5, 7.5), background=False,
...
-        else:
-            axes.set_title(title)
-        kwargs.pop("title")  # Do not give title to _finalise_figure
-    else:
-        plt.subplots_adjust(top=0.98)
     plt.tight_layout()
     plt.subplots_adjust(hspace=0)
-    fig = _finalise_figure(fig=fig, **kwargs)  # pragma: no cover
-    return fig
+    plt.subplots_adjust(top=0.93)
```
Comment (calum-chamberlain) [inline eqcorrscan/utils/plotting.py]: Please remove this logic - keep it in the calling function.

File: eqcorrscan/utils/plotting.py (hunk 1 above)
Comment (imankahbasi) [reply eqcorrscan/utils/plotting.py]: Ok.

**Comment by imankahbasi** [comment]
morning dear Chamberlian.
Thanks of you.
I test these changes with my own data that i work on it, and it works good! Then i thought it's good for pull request :( sorry.
Can i have some automatic test in my Github before make pull request?  (like yours) are these free?
i like to use kwargs, but i think it need some good and detailed description, because It may confuse users and they have to open source code and search for kwargs.
but is it possible to give **kwargs from function to function straightly? and make a good documentation for it.

**Comment by calum-chamberlain** [comment]
I've updated the plotting function docstrings so that all the supported kwargs are documented in PR #327, I will hopefully merge that into develop today and you should see the change.

The upshot is that `size` is a supported kwarg for all plotting functions now and is handled by `_finalise_figure`, so you shouldn't need to add that into the logic here.  I think you should just be able to get away with changes to `template_gen` functions.

**Comment by calum-chamberlain** [comment]
I updated your fork based on the current develop (with a standardised `size` kwarg, and docstrings). You will need to patch the template_plot calls as requested for this to work.

By the way, the section of the docs that I meant on testing is [here](https://eqcorrscan.readthedocs.io/en/latest/intro.html#runningtests).

**Comment by calum-chamberlain** [comment]
I'm hoping to get a new release out by the end of this week, it would be great to include this in it - if you can get it done overnight (for me) that would be awesome, otherwise I will just make some patches straight on develop including what you have here.  Thanks!

**Comment by imankahbasi** [comment]
> I'm hoping to get a new release out by the end of this week, it would be great to include this in it - if you can get it done overnight (for me) that would be awesome, otherwise I will just make some patches straight on develop including what you have here. Thanks!

morning!
I woke up now. and i just saw your message.
I try to finish it as soon as i can. But it might take 48 hours, because i don't have the adequate system for testing right now. But i am working on it and editing codes.
then i'll commit changes and see your automatic test here today.

**Comment by calum-chamberlain** [comment]
I;m having a go at this here.  I'm really not keen on putting all of this plotdir logic and figure resizing in hard coded into the plotting functions.  I wanted that logic to be in the functions that call them (e.g. `template_gen`), I'm also taking the opportunity to add a consistent plotdir argument to the match-filter objects.  I'll make a new PR based on my work in a few minutes (I'm just finishing running the tests) and you can see if that meets your needs.

**Comment by calum-chamberlain** [comment]
@imankahbasi have a look at #330 and check that that produces the plots that you want.

**Comment by imankahbasi** [comment]
Hi.
How can i find problem of code with continuous-tintegration/appveyor/pr?
I want to solve it but i don't know what is problem.

**Comment by calum-chamberlain** [comment]
It looks like appveyor just got stuck - this sometimes happens - network issues end up causing downloads to hang.  However, if #330 achieves the result you want then I would close this and merge that one in favour of this one.

**Comment by calum-chamberlain** [comment]
Closing in favour of #330

**Pull request closed without merging**